### ISSUE TITLE
Incorporate "./dev-install.sh" into "setup.py develop"?

### ISSUE THREAD

**Issue by mgeier** [jupyter-widgets/ipywidgets#2711]
... which would be automatically used when running

    python3 -m pip install -e .

This would make it much easier to get a development installation (assuming `yarn` is installed).

At some point, it might also make it possible to run a development install on RTD (but for now, `yarn` is not available there, at least not out-of-the-box).

Apart from that, the script `dev-install.sh` is unnecessarily platform-dependent, isn't it?

**Comment by jasongrout** [comment]
Is it possible to do this, but *not* do a dev-install when doing `pip install .`?

In my mind, dev-install does a pip install and more, so it doesn't make much sense to incorporate in to setup.py.

@vidartf also says dev-install doesn't work on windows, so I guess something is platform-dependent in there.

**Comment by jasongrout** [comment]
That said, if you can find a way to make it work, great. I think to do it right involves diving deep into the python packaging system and writing custom steps in the build process, but I'd love if it were simpler.

**Comment by mgeier** [comment]
> Is it possible to do this, but _not_ do a dev-install when doing `pip install .`?

Well, I guess that depends on what you mean by "dev-install".

Using `python3 -m pip install .` invokes `python3 setup.py install` (plus a few more details, see https://pip.pypa.io/en/stable/reference/pip_install/#build-system-interface).

I guess the real question is: What *should* `python3 setup.py install` do?

What's confusing to me (among other things) is how `widgetsnbextension` is handled.
Is it a separate project or not?
It seems to be a separate package on PyPI, yet it is somehow living in this repo ... what's up with that?

IIUC, when you currently run `python3 setup.py install` in the repo root directory, `widgetsnbextension` is installed from PyPI (as specified by `install_requires` in `setup.py`), regardless what the local directory `widgetsnbextension` contains (it might contain some new features?!).
However, when running `./dev-install.sh`, the local version of `widgetsnbextension` is installed (with `pip install -v -e .`).

That's, AFAIU, the current situation.
Now to how it probably should be ...

Before talking about "dev-install", we should probably talk about "source" vs. "binary" installation.

Most "normal" users will install the wheel package, which could be considered "binary" (`bdist_wheel`). A better word would probably be "pre-built", which means that "normal" users don't need to have `npm`, `yarn` and all this bulky stuff installed.

On the other side there is "source" installation, which itself can have 2 variations: installation from a tarball (`sdist`) or installation from the Git repo (be it from a local checkout or from a URL).
Since we are talking about a "source" installation, it makes sense to build the JavaScript stuff on demand.

Finally, when installing from source, you can choose whether you want an "editable" installation or not. In this case, the JavaScript stuff should be built as well. The difference is just that the files are not copied to some installation directory but instead linked in some way.

I would consider this last variation as a typical "dev-install".
But if you call it a "dev-install" whenever the JavaScript stuff is built, you would mean all the variations of "source" installation above.

Long story short, I would expect both `python3 setup.py install` and `python3 setup.py develop` to build the JavaScript stuff (since the JavaScript source files are part of the repo).

**Comment by jasongrout** [comment]
> What's confusing to me (among other things) is how `widgetsnbextension` is handled.
> Is it a separate project or not?
> It seems to be a separate package on PyPI, yet it is somehow living in this repo ... what's up with that?

It is a separate python package that lives in the same repo, and we do that so that it is easier to coordinate changes in the javascript packages.

> Long story short, I would expect both `python3 setup.py install` and `python3 setup.py develop` to build the JavaScript stuff (since the JavaScript source files are part of the repo).

A complication is that the top-level python package, `ipywidgets`, does *not* contain any javascript, so technically there's no reason to compile js when building and installing it. The js is part of the repo, but not part of the top-level `ipywidgets` python package.

dev-install builds an entire development build chain, including both the widgetsnbextension python package, as well as the jupyterlab npm module installed in JupyterLab.

I agree this is confusing. Part of that comes from widgets actually being two python packages and a number of interconnected js packages, one of which is a JupyterLab extension. To set up a full dev environment, you need to compile/install the whole set.

I'd love to make this more understandable, to lower the barrier of entry for people contributing.

**Comment by jasongrout** [comment]
It sounds like the discussion has basically run its course, except for the platform dependence of the dev install script. I'll leave this open for someone (who runs windows, presumably) to take a look at how to make the dev install work on Windows.

**Comment by mgeier** [comment]
Sorry @jasongrout, I wanted to look into this a bit more, but I didn't have the time ...

I understand that there can be some value in having different projects in the same repo to evolve them in lockstep, but is this really necessary here at this point in time?

I think separating the Python libraries from the JavaScript libraries (and putting each into their own repo) could make the whole system easier to understand. And it would most likely make the dev-install simpler.

And does `widgetsnbextension` even have to be a Python module/package?
It seems to contain very little Python code.

**Comment by jasongrout** [comment]
> is this really necessary here at this point in time?

I think it has more been a question of what made the workflow easiest for a developer. 

> I think separating the Python libraries from the JavaScript libraries (and putting each into their own repo) could make the whole system easier to understand. And it would most likely make the dev-install simpler.

Quite possibly, and it could make clearer the independence of the python vs js version numbers, which is something that is confusing in the current monorepo setup. It will take some nontrivial work to separate these, though. Would you like to look into doing this? We've already been separating the js libraries more for widgets 8.0.

> And does `widgetsnbextension` even have to be a Python module/package?

It's a python package because that is (used to be?) the easiest way to distribute a classic notebook extension. In fact, there have been suggestions that we make the JupyterLab manager a python package as well purely so that it is easier to distribute and install. If someone wants to work on a different proposal, however, happy to discuss this.

Perhaps at this point, if you have detailed thoughts about how you think the repo and packages could be better structured in a way that continues to make it easy to develop the widgets and is easier to understand, we'd love to hear a specific proposal (e.g., these js packages go into a this other repo like this, here's how the development process would work to keep versions in sync, here's what a typical change that involves both a python and js change would be developed, etc.). I don't have time to break apart this repo right now, but if someone wanted to put in the work and show how it made things better, I'm happy to help advise.

**Comment by maartenbreddels** [comment]
Fyi, I have a 'dummy' setup.py in vaex as well:
https://github.com/vaexio/vaex/blob/master/setup.py
I just have custom cmdclass to execute some custom code. Works in rtd, and
you don't have to explain to run some custom script (if it works)

(from mobile phone)

**Comment by jasongrout** [comment]
> Fyi, I have a 'dummy' setup.py in vaex as well:

Oh, interesting. So theoretically we could just execute `dev-install.sh` from setup.py when installing via `pip install -e`? It's sort of abusing the normal `pip install -e`, but perhaps it's worth it for the convenience?

**Comment by vidartf** [comment]
> So theoretically we could just execute dev-install.sh from setup.py when installing via pip install -e

Well, the script would have to be made platform independent, but otherwise I think that could work.

**Comment by mgeier** [comment]
I think the next step should be to move `widgetsnbextension` to a separate Github project.

This would already make the "dev-install" situation much simpler.

After that, we can continue with this issue.

**Comment by maartenbreddels** [comment]
Big fan of monorepo here, I prefer to keep related packages in 1 repo, much easier to organize refactors and multipackage PR's in general.

**Comment by mgeier** [comment]
A monorepo might have advantages (I don't know), but is it worth alienating potential contributors for that?

I guess this is, as so often, a matter of weighing different pros and cons.

I don't really have experience with monorepos, but isn't is essential to provide bespoke tools to handle them?
This lack of properly working tools is exactly the subject of this issue.

If running `python3 -m pip install -e .` would create a fully working dev installation, I guess I wouldn't really care whether `widgetsnbextension` is in a separate repo or not. I probably wouldn't even have noticed it in the first place.

**Comment by maartenbreddels** [comment]
Totally agree with you, a user should just do a pip install and it should 'just work'.

**Comment by jasongrout** [comment]
> I think the next step should be to move `widgetsnbextension` to a separate Github project.
> 
> This would already make the "dev-install" situation much simpler.

Just curious: exactly how do you think that will make things easier for a developer to have the widgetsnbextension package in a separate repo? This is an honest question - I'm not sure exactly what you gain by widgetsnbextension being a separate package from the rest of the js packages.

To me, the natural break in the packages is the kernel/server break. That would indicate that really the python ipywidgets package should be the one to split off, and keep all the frontend packages together so the js can be coordinated easier. That would mean that a developer that wanted to make a coordinated change between the kernel and frontend packages would need to do two dev installs (one for ipywidgets, one for the frontend package), but perhaps that is healthier to have that division very clear in your head.

**Comment by jasongrout** [comment]
> keep all the frontend packages together so the js can be coordinated easier.

This goes along with the ideas we had about having a `jupyter_widgets` python package with a variety of frontends packaged in it: https://github.com/jupyter-widgets/ipywidgets/pull/2807

**Comment by SylvainCorlay** [comment]
In the last weekly meeting, we discussed the idea of the separation of the Python ipywidgets package from the core.

At the moment, we feel that it may be a lot of work for little benefit. Theoretically, we *like* the idea of separating the base language agnostic package from the rest, but this may be too much work.

**Comment by mgeier** [comment]
First a disclaimer: I have no clue about JavaScript packaging. I'm looking at this purely from a Python point of view. So I might very well be missing something important.

@jasongrout 

> > I think the next step should be to move `widgetsnbextension` to a separate Github project.
> > This would already make the "dev-install" situation much simpler.
> 
> Just curious: exactly how do you think that will make things easier for a developer to have the widgetsnbextension package in a separate repo?

It would just be easier to understand.

Currently, the repo looks like it contains a Python package.

But strangely, it also contains a second, weirdly nested package. That's the first time I've seen such a thing in a Python project.

`widgetsnbextension` probably doesn't have to be a completely separate Github project if `ipywidgets` would be moved to a sub-directory at the same level as `widgetsnbextension`.
I still haven't seen something like this, but at least it would be easier to understand.

Apart from this being a highly unusual directory structure, it also made it very hard for me to understand what those packages even do.
And more importantly, I didn't know which package I should even care about. As far as I now understand, the packages seem to have actually very little overlap:

* `ipywidgets`: if you want to *create* widgets with IPython
  * Not relevant for languages other than Python.
  * Not relevant if you have pre-executed notebooks.
* `widgetsnbextension`: if you want to *view* widgets with the *Classic Notebook* (the widgets could have been created with any language)
  * Not relevant for JupyterLab
  * Not relevant for `nbsphinx`

Am I understanding this correctly so far?

Since `widgetsnbextension` is a sub-directory of `ipywidgets`, this was very hard to understand for me.

It just doesn't fit!

> This is an honest question - I'm not sure exactly what you gain by widgetsnbextension being a separate package from the rest of the js packages.

As I said, I don't know about the JS side, but this kind of nested Python packages is just weird.
And that's also the reason why it isn't possible to make a straightforward dev-install mechanism.
